### PR TITLE
Codify Infrastructure

### PR DESCRIPTION
**Pull request by caass**
Let's get this thing cracking.

[Link to RFC](https://github.com/caassandra/rfcs/blob/infra/text/0000-codify-infrastructure.md)

### CODE REVIEW

File: text/0000-codify-infrastructure.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+There's something to be said about rebuilding this setup to be less fragile, and there will likely be some issues trying to move this into Kubernetes. Kubernetes relies on its containers being mostly stateless, and so there may potentially be issues with having containers die in the middle of a stream.
+
+I mean, there shouldn't be, but there might be. Kubernetes is like that sometimes.
+
+Anyway, there are already docker images for both components of this setup, so wiring it up to Kubernetes should be mostly YAML configuring and not a lot of code editing.
+
+#### Video Streaming 📹
+
+Another cool think datafruits has is video streaming. The code for this lives (confusingly separate from streampusher) in [datafruits/viz](https://github.com/datafruits/viz). Currently, it's built on [`nginx-rtmp-module`](https://www.nginx.com/products/nginx/modules/rtmp-media-streaming/), which is a laughably simple way to set up video streaming. The repo reflects this, having literally two files.
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:153]:
```suggestion
Another cool thing that datafruits has is video streaming. The code for this lives (confusingly separate from streampusher) in [datafruits/viz](https://github.com/datafruits/viz). Currently, it's built on [`nginx-rtmp-module`](https://www.nginx.com/products/nginx/modules/rtmp-media-streaming/), which is a laughably simple way to set up video streaming. The repo reflects this, having literally two files.
```

**Comment by mcfiredrill** [comment]
Will there be any issue running liquidsoap's harbor input through the proxy, or will that be run through the proxy at all?
DJs stream using one of these desktop clients to an endpoint on liquidsoap called the harbor.
http://mcfiredrill.github.io/blog/big-list-of-source-clients

It's not exactly standard HTTP requests, so I'm not sure if there will be any issues. Something to think about.

**Comment by caassandra** [comment]
@mcfiredrill i'm pretty sure kubernetes ingress can handle any requests made over TCP (or UDP, if we really wanted), so i think streaming audio should be fine. i agree though, worth looking into further.

File: text/0000-codify-infrastructure.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+
+This is the place you go where if you're Tony, you can look at all the DJs and their information, or if you're a DJ you can manage your own stuff and upload sets and things like that. I'm not _super_ sure how it works, because I'm not a DJ and I don't have access to it, and I've never been invested enough to really dig through the code.
+
+That said, it has its entire own codebase, which lives in [streampusher/frontend](https://github.com/streampusher/frontend). This, too, is an Ember app, which does NOT use ember-cli-fastboot. It currently lives on [netlify](https://www.netlify.com/), which is like a [serverless](https://aws.amazon.com/serverless/) [static site](https://en.wikipedia.org/wiki/Static_web_page) hosting platform, kind of like [GitHub Pages](https://pages.github.com/). I think. I've never used netlify, although I certainly have seen people sing its praises.
+
+At least, I think it lives on netfliy. There's a netlify config in the [`ember-cli-build.js`](https://github.com/streampusher/frontend/blob/master/ember-cli-build.js). Either way, it should be similarly easy to containerize this app the same way we'd containerize the client site.
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:83]: I don't think I have plans to ever convert the streampusher frontend to fastboot...there is nothing really you can see on the site unless you login, so there's no concerns for SEO or anything.

Would a static site really benefit from being containerized? It seems like it would be more work in a way, we'd have to setup some kind of server to get it running. 

Is there some other way to serve a static site that is more in line with your new infrastructure proposal? Netlify is super simple to use right now, and while I'm not opposed to using something else, all I need to do is push to master to deploy now, and its up in a minute, and I'd like to keep that kind of workflow if possible.

Comment (caassandra) [reply text/0000-codify-infrastructure.md:83]: yea absolutely! netlify is basically a static file server, so we could just set up CI to deploy the files to object storage and configure the DNS to point to it, digitalocean would be basically a drop-in replacement

File: text/0000-codify-infrastructure.md [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+
+The solution to this is the new shiny thing called [GraphQL](https://graphql.org/), which aims to solve the problems that REST has. Refactoring the existing API into GraphQL is gonna probably warrant its own RFC, but I wanted to mention it here so I didn't forget about it.
+
+In terms of containerization, the API literally [already has a Dockerfile](https://github.com/streampusher/api/blob/master/Dockerfile) so this should be ok with only minor tweaks.
+
+#### Audio streaming 🔊
+
+This is probably the second most important part of the ᴅᴀᴛᴀғʀᴜɪᴛs ᴇxᴘᴇʀɪᴇɴᴄᴇ, behind the site itself. The code that handles this stuff lives in [streampusher/radio](https://github.com/streampusher/radio). The audio streaming is built on [icecast](https://icecast.org), which is what broadcasts the audio, and [liquidsoap](https://www.liquidsoap.info/), which provides icecast with the audio to stream.
+
+There's something to be said about rebuilding this setup to be less fragile, and there will likely be some issues trying to move this into Kubernetes. Kubernetes relies on its containers being mostly stateless, and so there may potentially be issues with having containers die in the middle of a stream.
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:145]: Can we talk about what is fragile with the current setup? Don't hold back. :)

I feel like the main problem is DJs disconnecting due to them being on wifi instead of ethernet, and that's not exactly something we can fix on our side I think. 

Also sometimes the cert for the relay expires and letsencrypt doesn't seem to renew it automatically, that's definitely something we could fix.

Anything else?

File: text/0000-codify-infrastructure.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+
+At least, I think it lives on netfliy. There's a netlify config in the [`ember-cli-build.js`](https://github.com/streampusher/frontend/blob/master/ember-cli-build.js). Either way, it should be similarly easy to containerize this app the same way we'd containerize the client site.
+
+#### Static Assets 🗂
+
+Another important thing for UX is the static assets -- who doesn't love the pulsating 𝕕𝕒𝕥𝕒𝕗𝕣𝕦𝕚𝕥𝕤 logo? Currently...I'm not sure how these are served. Some of them are checked in to the client site repo, but some of them aren't. With this RFC, I propose we move all of the static assets to dedicated object hosting (like [DigitalOcean Spaces](https://www.digitalocean.com/products/spaces/)). [More on object storage](https://blog.westerndigital.com/why-object-storage/)
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:87]: I would love to move all the podcast mp3s to spaces. :tada: 

Also get a real CDN going. I'm not sure what our options are there. I've heard good things about [bunny cdn](https://bunnycdn.com/).

File: text/0000-codify-infrastructure.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+
+We could minimize the pain of uploading assets separately and harcoding urls by editing the config for [`broccoli-asset-rev`](https://github.com/rickharrison/broccoli-asset-rev) (the tooling ember uses to process static assets) to prepend a url to the assets in production, and then just check everything in to the client site repo under `src/assets/`.
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:89]: Not `public/assets/`  ?

File: text/0000-codify-infrastructure.md [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+
+This also allows us to move from treating the services (streampusher, datafruits) like pets to treating them like cattle, which means less messy ssh-ing and more codified changes. [More reading on this terminology](https://medium.com/@Joachim8675309/devops-concepts-pets-vs-cattle-2380b5aab313)
+
+Additionally, by moving to a [Kubernetes (K8s)](https://kubernetes.io/) based infrastructure, it allows us a lot more (horizontal) scalability, and the ability to scale out each chunk of the stack individually -- if there are 30,000,000 users in the chat, it makes sense to scale up the chat server, but maybe not also the [rmtp(s)](https://en.wikipedia.org/wiki/Real-Time_Messaging_Protocol) server, which only has the one person running viz connected to it.
+
+Fourth, we now have the opportunity to actually do logging. Logging! Imagine, datafruits goes down and we don't have to spend eight hours figuring out why. We have an idea for a new feature, and we can actually figure out if people use it. Metrics! God, the metrics. Plus we can avoid G\*\*\*\*\* A\*\*\*\*\*\*\*\* and run our [own analytics platform](https://github.com/zgoat/goatcounter) that's privacy friendly.
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:59]: Can you elaborate more on how the logging would work? Is there some kind of web interface? I'm new to this...

File: text/0000-codify-infrastructure.md [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+
+Lastly, there are also cost savings associated with moving to a containerized architecture. We move from running a few digital ocean droplets, a couple heroku dynos, and whatever else we have to running 3 droplets on digital ocean. Additionally, those droplets will be a lot more efficient about their resource usage than they currently are, since their only purpose will be to run docker images.
```
Comment (mcfiredrill) [inline text/0000-codify-infrastructure.md:61]: Current costs are something like
1 digital ocean droplet for streampusher api/postgresql/redis/liquidsoap/main icecast -> $40
1 digital ocean droplet for icecast relay -> $5
heroku dyno for datafruits.fm w/ fastboot -> $7
heroku dyno for hotdoglounge chat -> $7

I think that's all...

File: text/0000-codify-infrastructure.md (hunk 3 above)
Comment (caassandra) [reply text/0000-codify-infrastructure.md:145]: > Can we talk about what is fragile with the current setup? Don't hold back. :)

Ahhh lol we'll I'm by *no* means an expert on this stuff. I mostly mean "fragile" in terms of "it's hard to understand what changes will break things." Similarly to how it would be nice to have all of the infrastructure documented, it would be nice to have some more idea about how all the different files in the repo fit together.

> I feel like the main problem is DJs disconnecting due to them being on wifi instead of ethernet, and that's not exactly something we can fix on our side I think.

I agree, and it might be worth tinkering with the [timeout](https://www.liquidsoap.info/doc-dev/harbor.html#parameters).

> Also sometimes the cert for the relay expires and letsencrypt doesn't seem to renew it automatically, that's definitely something we could fix.

Yeah, in our cluster we'd have a dedicated [certificate manager](https://cert-manager.io/docs/) to make sure all our stuff is encrypted, so we'd offload the security functionality to there. This also I think is a benefit because it decouples security from radio.

> Anything else?

Off the top of my head:

- There's three top level files ([start.sh](https://github.com/streampusher/radio/blob/master/start.sh), [setup.sh](https://github.com/streampusher/radio/blob/master/setup.sh), and a [Vagrantfile](https://github.com/streampusher/radio/blob/master/Vagrantfile)) which all seem to serve a similar purpose, except all three of them haven't been touched in years.
- Similarly, there's a *bunch* of files in [/liquidsoap](https://github.com/streampusher/radio/tree/master/liquidsoap) that haven't been touched in years. Some Ruby, some Bash, some config, etc. I'd have a lot more confidence if I could look at the files and know that they're untouched because they're perfect, and not because they're forgotten.
- More generally, I think it makes sense to set up a docker-compose.yml or something in the top-level directory indicating how the different images work together. Unfortunately Kubernetes doesn't play *super* well with docker-compose but there's tools like [Kompose](https://kubernetes.io/docs/tasks/configure-pod-container/translate-compose-kubernetes) that we could drop into CI to auto-magically update the Kubernetes config while also allowing redistribution of the standalone radio stuff.

File: text/0000-codify-infrastructure.md (hunk 4 above)
Comment (caassandra) [reply text/0000-codify-infrastructure.md:87]: DigitalOcean has a [built-in CDN](https://www.digitalocean.com/products/spaces/#testContentBlockSection) for their delivery :)

Also, if your DNS is with [Cloudflare](https://www.cloudflare.com/learning/cdn/what-is-caching/), they have built-in caching too.

Btw, if you're not with Cloudflare, i highly recommend them. They recently started offering zero-markup domain names, and their free plan is super generous.

File: text/0000-codify-infrastructure.md (hunk 5 above)
Comment (caassandra) [reply text/0000-codify-infrastructure.md:89]: Yes :) lol

File: text/0000-codify-infrastructure.md (hunk 6 above)
Comment (caassandra) [reply text/0000-codify-infrastructure.md:59]: Yeah sure! Distributed logging is actually a kind of big field, because everything is so spread out (heh). Here's a [quick explanation](https://www.reddit.com/r/devops/comments/9hku3v/prometheus_vs_opentracing/e6db02o) of "observability", which is CS-speak for understanding what's going on inside a distributed system. Copied here for your convenience:

> There are 3 pillars of observability:
> 
> * metrics
> * logs
> * tracing
> 
> While all of them partially overlap, each have different purpose.
>
> #### Metrics
> 
> These are numbers related amount of events that happened in the time range. Like:
> 
> * number of successful/failed/overall requests
> * cumulative duration of requests
> * bucketed histogram of requests’ durations
> 
> The main point of metrics is that these are small, you can gather them cheaply and store them for a long period of time. These give you overall overview of the whole system, but without insights.
> 
> So, metrics answer to the question “how is my system performance changes through time?”
>
> #### Logs
> 
> These are single events that happened in the system, ex. single request. It will often contain also some informations that are in the metrics (like request duration), but it will also contain more context, like IP of the requester and precise time of the request. The problem with logs isn’t that these are larger in size than just metrics itself, so you cannot store them for so long as the metrics and due tho their size, you need to reduce amount of logs you send.
> 
> So, logs answers to the question “what happened in my system?”.
>
> #### Tracing
> 
> This one is for monitoring how single event is behaving in the system. So you have information how long this request spent in LB, backend, DB, how long it spend in external services, and so on. So as you see, this overlap with both of metrics and logs, but still not exactly. The main problem is that while metrics and logs can have different formats in each of the services, tracing needs to be uniform, because your system needs to be “request ID aware”. Also traces can be very large, that is why in general you do not trace all requests, but only some of them to reduce internal traffic.
> 
> So, tracing answers to the question “how is my system components interacting with each other?”.

#### Cluster-Level Logging

This is like, the classic version of logging. You have an application, and it spits some stuff out via `console.log` or `console.error`, and you want to capture that. In Kubernetes, every container's `stdout` and `stderr` is captured, and you have the option of monitoring logs either at the Pod (container group) level or at the Node (host VM) level.

We'd almost certainly do it at the Node level, because that doesn't add any additional overhead to application programming. Here's a nice image [from the docs](https://kubernetes.io/docs/concepts/cluster-administration/logging/#using-a-node-logging-agent):

node logging

What you're looking at is a simple logging configuration for one node. You have your application container, which streams to `stdout` and `stderr`, which is then redirected by Kubernetes to a log file on the Node. The "logging agent", which is actually called [Fluentd](https://docs.fluentd.org/), parses these logs and sends them off to some other log aggregation service. `logrotate` is a built-in Kubernetes tool that just deletes old logs once they take up a certain amount of disk space.

#### Application-Level Logging

This is a little more involved than just streaming to `stdout` and `stderr`, but not much more so. At least, on the application side. Usually what you do is use something like a [Prometheus client library](https://github.com/deadtrickster/prometheus.ex) and write something like this:

```ex
defmodule ExampleInstrumenter do
  use Prometheus.Metric

  @histogram [name: :http_request_duration_milliseconds,
              labels: [:method],
              buckets: [100, 300, 500, 750, 1000],
              help: "Http Request execution time"]

  def instrument(%{time: time, method: method}) do
    Histogram.observe([name: :http_request_duration_milliseconds, labels: [method]], time)
  end
end
```

This is useful to monitor metrics from within individual applications, such as response times on the API or which DJs are the most obsessive about updating their DF profiles. It's basically strictly better than cluster-level logging, except that it requires you to actually write code dedicated to instrumenting logging instead of just `console.log`-ing.

#### Cluster Health

I know. Why is this not the thing called "cluster-level logging"? The answer is that Google is bad at naming things. This refers to monitoring things like CPU usage, available disk space, that sort of thing. There's a [full writeup here](https://phoenixnap.com/kb/prometheus-kubernetes-monitoring), but basically it just involves pointing your log aggregator at the built-in Kubernetes monitoring endpoints to keep tabs on the Nodes and the code that's running the cluster itself. They included this nice picture.

cluster health

#### Generating metrics from logs

As you may have guessed, the go-to distributed logging tool is [Prometheus](https://prometheus.io/). The idea is that you can plug a bunch of data sources into it, and then it performs some log aggregation before storing results in a [time series database](https://www.influxdata.com/time-series-database/). It can be configured to delete old logs to avoid paying too much for storage, and also to delete *most* of the old logs to allow for checking metrics -- for example, you don't need to know the status of all 2,000 requests to the API from January 23, but it would be nice to know that the average response time was 1.5 seconds.

#### Tracing

Tracing is about tracking a single event as it moves through your distributed system. For example, think about what happens when a DJ wants to upload a new set.

1. DJ clicks "upload mp3" on streampusher, sending a POST request to api.datafruits.com
3. The ingress controller routes the request to the API
3. The API authenticates the request and forwards the image to object storage
4. Object storage stores the file and responds 200 OK
5. The API receives the response from object storage and responds 200 OK
6. The ingress controller forwards the response to the DJ, and streampusher says "Success!"

Even though this is one event, it touches the ingress controller, the API, and object storage. If we're trying to troubleshoot why it takes a million years to upload a set, we need to be able to trace this event. That's where something like [Jaeger](https://www.jaegertracing.io/docs/1.17/#about) comes in. This gets a little more involved, because it requires instrumenting all the code in the event lifecycle, but eventually it'd be nice to have.

#### Monitoring

Prometheus can do automated alerts (via a bunch of stuff, including [discord](https://github.com/prometheus/alertmanager/issues/1365#issuecomment-579337769)), and we can also use tools like [Grafana](https://prometheus.io/docs/visualization/grafana/) to build dashboards.

#### TL;DR

Basically, we need some kind of aggregation service to centralize our logs, and a visualization tool to make a dashboard to look at those logs.

File: text/0000-codify-infrastructure.md (hunk 7 above)
Comment (caassandra) [reply text/0000-codify-infrastructure.md:61]: Just using some quick napkin math based on [Digital Ocean pricing](https://www.digitalocean.com/pricing/):

[Managed Postgres DB](https://www.digitalocean.com/pricing/#managed-databases):

Memory | vCPUs | Disk | Standby Nodes | $/HR | $/MO
-- | -- | -- | -- | -- | -- |
1GB | 1vCPU | 10GB | N/A | $0.022 | $15

[Spaces Object Storage](https://www.digitalocean.com/pricing/#spaces-object-storage):

Storage | Outbound Transfer | Add'l. GB Stored | Add'l. GB Transferred | $/MO 
-- | -- | -- | -- | --
250GB | 1TB | $0.02/GB | $0.01/GB | $5

[Standard Droplets](https://www.digitalocean.com/pricing/#standard-droplets):

Memory | vCPUs | Transfer | SSD Disk | $/HR | $/MO
-- | -- | -- | -- | -- | -- 
1GB | 1vCPU | 1TB | 25GB | $0.007 | $5

Just doing some quick napkin math:

Component | $/MO
-- | --
Droplet (x3) | $15
Postgres | $15
Object Storage | $5
**Total** | **$35**

I didn't realize we were using redis -- what's that being used for?

Comment (mcfiredrill) [reply text/0000-codify-infrastructure.md:61]: > I didn't realize we were using redis -- what's that being used for?

Sidekiq (background worker) process in Rails, some other things in Rails, its also a way for Rails to communicate with the hotdog lounge chat app.

File: text/0000-codify-infrastructure.md (hunk 2 above)
Comment (mcfiredrill) [reply text/0000-codify-infrastructure.md:83]: Right, I used to do this with S3. DO Spaces isn't exactly a CDN though, afaik, is there something more sophisticated CDN-like we could use?

Comment (caassandra) [reply text/0000-codify-infrastructure.md:83]: Right, it's a little more involved than just pointing it at a bucket -- [here](https://www.digitalocean.com/docs/spaces/#cdn) they talk about how to set it up. Basically, you replace <spacename>.<region>.digitaloceanspaces.com with <spacename>.<region>.*cdn*.digitaloceanspaces.com to activate it, and then you can [assign a custom subdomain](https://www.digitalocean.com/docs/spaces/how-to/customize-cdn-endpoint/) if you've registered the domain name with digital ocean.

File: text/0000-codify-infrastructure.md (hunk 7 above)
Comment (caassandra) [reply text/0000-codify-infrastructure.md:61]: Gotcha -- If that's the case we'd likely use what's called the [sidecar pattern](https://thenewstack.io/tutorial-apply-the-sidecar-pattern-to-deploy-redis-in-kubernetes/) where we'd have a Redis client container alongside both the hotdog_lounge containers and the Rails API container, and both of those would connect to a [managed redis instance](https://www.digitalocean.com/products/managed-databases-redis/). Adding that on would bring the total to $50/MO.

The sidecar pattern is also nice because for local development you can still use localhost, since the containers will share a pod.

File: text/0000-codify-infrastructure.md (hunk 3 above)
Comment (mcfiredrill) [reply text/0000-codify-infrastructure.md:145]: > There's three top level files (start.sh, setup.sh, and a Vagrantfile) which all seem to serve a similar purpose, except all three of them haven't been touched in years.
Similarly, there's a bunch of files in /liquidsoap that haven't been touched in years. Some Ruby, some Bash, some config, etc. I'd have a lot more confidence if I could look at the files and know that they're untouched because they're perfect, and not because they're forgotten.

Sorry about all that, totally my fault. I will try to clean them up a bit.

Comment (mcfiredrill) [reply text/0000-codify-infrastructure.md:145]: > More generally, I think it makes sense to set up a docker-compose.yml or something in the top-level directory

There is one in [streampusher/api](https://github.com/streampusher/api/blob/master/docker-compose.yml) , were you thinking its not enough?

Comment (caassandra) [reply text/0000-codify-infrastructure.md:145]: I appreciate it! And no worries, I totally understand getting it up and running and going "ok done with this forever" lol

**Comment by mcfiredrill** [comment]
By the way the Rails app is deployed currently with capistrano, would that still be necessary or is there another way to deploy with this setup?

**Comment by caassandra** [comment]
@mcfiredrill i'm not familiar with capistrano but from my quick 30 second glance it seems like a cd tool, which would be replaced by the same workflow of all the other app repositories that follow the docker workflow:

1. code is pushed to a repo
2. ci runs tests
3. ci builds & pushes docker image
4. kubernetes detects new image
5. kubernetes spins up new version
6. kubernetes kills old version

the db connectivity aspects of capistrano will also be unnecessary because with the sidecar pattern, as far as our rails app is concerned it's just connecting to some port on localhost

File: text/0000-codify-infrastructure.md (hunk 6 above)
Comment (mcfiredrill) [reply text/0000-codify-infrastructure.md:59]: Wow this is certainly a lot bigger answer than I expected! :sweat_smile: 
What would an MVP of this look like, and is it still within the scope of this RFC?

**Comment by mcfiredrill** [comment]
Another thing I'd like to do in the future is look into the possibility of adding multiple relays for different geographic areas to better serve listeners....I imagine once we have this setup going that kind of thing would be easier?

**Comment by caassandra** [comment]
@mcfiredrill hmm i think so -- from what i remember in the icecast documentation the relays just need an IP address to forward the streams from, and so i suspect that we could just feed them the address of the cluster ingress and route those requests to the broadcast service.

**Comment by mcfiredrill** [comment]
Unless there's anything blocking this I think I will merge this week or next, if anyone has any more comments, now its a great time to write them!

I'm going to make a datafruits/infra repo for @caassandra to get started in, and we will create issues and a project roadmap from there.

**Comment by caassandra** [comment]
Should we put a number to this RFC instead of 0000? Idk how that all works.

**Comment by caassandra** [comment]
Seeing as we now have [datafruits/infra](https://github.com/datafruits/infra) I'm gonna be moving planning / roadmap / etc. over there 🙃
